javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -0,0 +1,737 @@
+
+        /*
+       * fun deleteMin [] = raise EMPTY
+       * * | deleteMin ts =
+       * *     let fun getMin [t]=(t, [])
+       * *           | getMin (t :: ts) =
+       * *               let val (t', ts') = getMin ts
+       * *               in if Elem.leq(root t, root t') then (t, ts)
+       * *                  else                              (t', t :: ts') end
+       * *         fun split (ts,xs,[]) = (ts, xs)
+       * *           | split (ts,xs,t :: c) =
+       * *               if rank t = 0 then split (ts,root t :: xs,c)
+       * *               else               split (t :: ts,xs,c)
+       * *         val (Node (x,r,c), ts) = getMin ts
+       * *         val (ts',xs') = split ([],[],c)
+       * * in fold insert xs' (meld (ts, ts')) end
+       * */
+        protected static <T> List<Node<Ranked<T>>> deleteMin(SerializableComparator<? super T> comparator, List<Node<Ranked<T>>> forest) {
+            if (forest.isEmpty()) {
+                throw new NoSuchElementException();
+            } else {
+                /* get the minimum tree and the rest of the forest */
+                final Node<Ranked<T>> minTree = forest.reduce((t1, t2) -> comparator.isLessOrEqual(root(t1), root(t2)) ? t1 : t2); // TODO could we use PriorityQueue here instead ... no, really
+                final List<Node<Ranked<T>>> forestTail = (minTree == forest.head()) ? forest.tail() : forest.remove(minTree);
+
+                final List<Node<Ranked<T>>> newForest = rebuild(comparator, minTree.getChildren());
+                return meld(comparator, newForest, forestTail);
+            }
+        }
+
+        /* Separate the rank 0 trees from the rest, rebuild the 0 rank ones and merge them back */
+        private static <T> List<Node<Ranked<T>>> rebuild(SerializableComparator<? super T> comparator, List<Node<Ranked<T>>> forest) {
+            List<Node<Ranked<T>>> nonZeroRank = List.empty(), zeroRank = List.empty();
+            for (; !forest.isEmpty(); forest = forest.tail()) {
+                final Node<Ranked<T>> initialForestHead = forest.head();
+                if (rank(initialForestHead) == 0) {
+                    zeroRank = insert(comparator, zeroRank, root(initialForestHead));
+                } else {
+                    nonZeroRank = concat(initialForestHead, nonZeroRank);
+                }
+            }
+            return meld(comparator, nonZeroRank, zeroRank);
+        }
+
+        /**
+         * fun insert (x, ts as t1 :: t2 :: rest) =
+         * *     if rank t1 = rank t2 then skewLink(Node(x,0,[]),t1,t2) :: rest
+         * *     else                      Node (x,0,[]) :: ts
+         * * | insert (x, ts) =            Node (x,0,[]) :: ts
+         **/
+        protected static <T> List<Node<Ranked<T>>> insert(SerializableComparator<? super T> comparator, List<Node<Ranked<T>>> forest, T element) {
+            Node<Ranked<T>> tree = Tree.of(Ranked.of(element, 0));
+            if (forest.size() >= 3) {
```

Comment:
oups, this was faulty, I presumed that `t1 :: t2 :: rest` meant 3 elements, though rest can also be an empty list, of course :)